types/src/state_store/state_key_prefix.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::state_store::state_key::{StateKey, StateKeyTag};
+use std::iter::zip;
+
+// Struct for defining prefix of a state key, which can be used for finding all the values with a
+// particular key prefix
+#[derive(Debug)]
+pub struct StateKeyPrefix {
+    tag: StateKeyTag,
+    bytes: Vec<u8>,
+}
+
+impl StateKeyPrefix {
+    pub fn new(tag: StateKeyTag, bytes: Vec<u8>) -> Self {
+        Self { tag, bytes }
+    }
+
+    /// Serializes to bytes for physical storage.
+    pub fn encode(&self) -> anyhow::Result<Vec<u8>> {
+        let mut out = vec![self.tag.clone() as u8];
+        out.extend(self.bytes.clone());
+        Ok(out)
+    }
+
+    /// Checks if the current prefix is a valid prefix of a particular state_key
+    pub fn is_prefix(&self, state_key: &StateKey) -> anyhow::Result<bool> {
+        let encoded_key = state_key.encode()?;
+        let encoded_prefix = self.encode()?;
+        // Check if bytes is a sub-vector of encoded key.
+        if encoded_prefix.len() > encoded_key.len() {
+            return Ok(false);
+        }
+        for (x, y) in zip(encoded_prefix, encoded_key) {
+            if x != y {
+                return Ok(false);
+            }
+        }
+        Ok(true)
```

Comment:
compare with `encoded_key[..encode_prefix.len()]`? we probably don't need a manual loop here.

Reply:
I don't know what I was thinking :D